server/controllers/message.js

```diff
@@ -0,0 +1,164 @@
+const asyncHandler = require('express-async-handler');
+const decodeToken = require('../utils/decodeToken');
+
+const Message = require('../models/Message');
+const Conversation = require('../models/Conversation');
+
+// @route POST /conversation/create/:id
+// Create a conversation
+exports.newConversation = asyncHandler(async (req, res) => {
+  const recipient = req.params.id;
+  const { type, content } = req.body;
+  let decoded = decodeToken(req.cookies.token);
+  const userId = decoded.id;
+  try {
+    const existingConversation = await Conversation.find({
+      $and: [{ participants: { $in: userId } }, { participants: { $in: recipient } }, { deleted: false }],
+    });
+    if (existingConversation.length > 0) {
+      res.status(403).json({ error: 'Conversation already exists' });
+    } else {
+      const message = await Message.create({
+        sender: userId,
+        recipient,
+        sendDate: Date.now(),
+        content,
+        type,
+      });
+      await Conversation.create({
+        participants: [userId, recipient],
+        messages: [message],
+        mostRecentMsg: message,
+        unreadMsgs: 1,
+      });
+      res.status(200).json({
+        success: {
+          message: {
+            id: message._id,
+            content: message.content,
+            sendDate: message.sendDate,
+            read: message.read,
+          },
+        },
+      });
+    }
+  } catch (error) {
+    return res.status(500).json({ error: 'Could not create conversation' });
+  }
+});
+
+// @route POST /:id
+// Create a new message
+exports.newMessage = asyncHandler(async (req, res) => {
+  const { type, content } = req.body;
+  const recipient = req.params.id;
+  let decoded = decodeToken(req.cookies.token);
+  const userId = decoded.id;
+  try {
+    const message = await Message.create({
+      sender: userId,
+      recipient,
+      sendDate: Date.now(),
+      content,
+      type,
+    });
+    await Conversation.findOneAndUpdate(
+      { $and: [{ participants: { $in: userId } }, { participants: { $in: recipient } }] },
+      { $push: { messages: message }, mostRecentMsg: message, $inc: { unReadMsgs: 1 } },
```

Comment:
We need to only push the message's id under messages, I don't think we need to push the entire message here :)